File: nextjs/lib/utils/sanitization.ts
/**
 * Input sanitization utilities for client-side validation and sanitization.
 */

/**
 * Sanitize a string by stripping whitespace and validating length.
 *
 * @param value - String to sanitize
 * @param maxLength - Maximum allowed length (undefined for no limit)
 * @param minLength - Minimum required length
 * @returns Sanitized string
 * @throws Error if string is empty after stripping or doesn't meet length requirements
 */
export function sanitizeString(value: string, maxLength?: number, minLength: number = 0): string {
  if (typeof value !== "string") {
    throw new Error("Value must be a string");
  }

  const sanitized = value.trim();

  if (sanitized.length < minLength) {
    throw new Error(`String must be at least ${minLength} characters long`);
  }

  if (maxLength !== undefined && sanitized.length > maxLength) {
    throw new Error(`String must be at most ${maxLength} characters long`);
  }

  return sanitized;
}

/**
 * Validate and sanitize an email address.
 *
 * @param email - Email address to validate
 * @returns Sanitized email address
 * @throws Error if email is invalid
 */
export function sanitizeEmail(email: string): string {
  if (typeof email !== "string") {
    throw new Error("Email must be a string");
  }

  const sanitized = email.trim().toLowerCase();

  // Basic email validation regex
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  if (!emailRegex.test(sanitized)) {
    throw new Error("Invalid email format");
  }

  return sanitized;
}

/**
 * Validate and sanitize a URL.
 *
 * @param url - URL to validate
 * @param allowedSchemes - Array of allowed URL schemes (default: ['http', 'https'])
 * @returns Sanitized URL
 * @throws Error if URL is invalid or uses a disallowed scheme
 */
export function sanitizeUrl(url: string, allowedSchemes: string[] = ["http", "https"]): string {
  if (typeof url !== "string") {
    throw new Error("URL must be a string");
  }

  const sanitized = url.trim();

  if (!sanitized) {
    throw new Error("URL cannot be empty");
  }

  try {
    const parsed = new URL(sanitized);

    if (!parsed.protocol) {
      throw new Error("URL must include a scheme (e.g., http:// or https://)");
    }

    const scheme = parsed.protocol.replace(":", "").toLowerCase();
    if (!allowedSchemes.map((s) => s.toLowerCase()).includes(scheme)) {
      throw new Error(`URL scheme must be one of: ${allowedSchemes.join(", ")}`);
    }

    return sanitized;
  } catch (error) {
    if (error instanceof TypeError) {
      throw new Error("Invalid URL format");
    }
    throw error;
  }
}

/**
 * Sanitize a slug by removing invalid characters and ensuring proper format.
 *
 * @param slug - Slug to sanitize
 * @returns Sanitized slug
 */
export function sanitizeSlug(slug: string): string {
  if (typeof slug !== "string") {
    throw new Error("Slug must be a string");
  }

  // Convert to lowercase and replace spaces with hyphens
  let sanitized = slug.trim().toLowerCase().replace(/\s+/g, "-");

  // Remove invalid characters (only allow alphanumeric and hyphens)
  sanitized = sanitized.replace(/[^a-z0-9-]/g, "");

  // Remove consecutive hyphens
  sanitized = sanitized.replace(/-+/g, "-");

  // Remove leading and trailing hyphens
  sanitized = sanitized.replace(/^-+|-+$/g, "");

  return sanitized;
}

File: backend/utils/sanitization.py
"""Input sanitization utilities."""

import re
from urllib.parse import urlparse

from core.constants import ValidationErrorMessages


def sanitize_string(value: str, max_length: int | None = None, min_length: int = 0) -> str:
    """
    Sanitize a string by stripping whitespace and validating length.

    Args:
        value: String to sanitize
        max_length: Maximum allowed length (None for no limit)
        min_length: Minimum required length

    Returns:
        Sanitized string

    Raises:
        TypeError: If value is not a string
        ValueError: If string is empty after stripping or doesn't meet length requirements
    """
    if not isinstance(value, str):
        raise TypeError(ValidationErrorMessages.VALUE_MUST_BE_STRING)

    sanitized = value.strip()

    if len(sanitized) < min_length:
        raise ValueError(ValidationErrorMessages.STRING_TOO_SHORT.format(min_length=min_length))

    if max_length is not None and len(sanitized) > max_length:
        raise ValueError(ValidationErrorMessages.STRING_TOO_LONG.format(max_length=max_length))

    return sanitized


def sanitize_filename(filename: str) -> str:
    """
    Sanitize a filename by removing dangerous characters.

    Args:
        filename: Filename to sanitize

    Returns:
        Sanitized filename

    Raises:
        TypeError: If filename is not a string
        ValueError: If filename is empty after sanitization
    """
    if not isinstance(filename, str):
        raise TypeError(ValidationErrorMessages.FILENAME_MUST_BE_STRING)

    # Remove path separators and other dangerous characters
    sanitized = re.sub(r'[<>:"/\\|?*\x00-\x1f]', "", filename)
    sanitized = sanitized.strip()

    if not sanitized:
        raise ValueError(ValidationErrorMessages.FILENAME_CANNOT_BE_EMPTY)

    # Limit length
    if len(sanitized) > 255:
        sanitized = sanitized[:255]

    return sanitized


def sanitize_url(url: str, allowed_schemes: list[str] | None = None) -> str:
    """
    Validate and sanitize a URL.

    Args:
        url: URL to sanitize
        allowed_schemes: List of allowed URL schemes (default: ['http', 'https'])

    Returns:
        Sanitized URL

    Raises:
        TypeError: If URL is not a string
        ValueError: If URL is invalid or uses a disallowed scheme
    """
    if not isinstance(url, str):
        raise TypeError(ValidationErrorMessages.URL_MUST_BE_STRING)

    sanitized = url.strip()

    if not sanitized:
        raise ValueError(ValidationErrorMessages.URL_CANNOT_BE_EMPTY)

    try:
        parsed = urlparse(sanitized)
    except Exception as e:
        raise ValueError(ValidationErrorMessages.URL_INVALID_FORMAT.format(error=str(e))) from e

    if not parsed.scheme:
        raise ValueError(ValidationErrorMessages.URL_MISSING_SCHEME)

    if allowed_schemes is None:
        allowed_schemes = ["http", "https"]

    if parsed.scheme.lower() not in [s.lower() for s in allowed_schemes]:
        raise ValueError(
            ValidationErrorMessages.URL_INVALID_SCHEME.format(schemes=", ".join(allowed_schemes))
        )

    return sanitized


def sanitize_html(html: str, allowed_tags: list[str] | None = None) -> str:
    """
    Basic HTML sanitization by removing HTML tags.

    Args:
        html: HTML string to sanitize
        allowed_tags: List of allowed HTML tags (None means strip all tags)

    Returns:
        Sanitized string with HTML tags removed

    Note:
        This is a basic implementation. For production use, consider using
        a library like bleach for more comprehensive HTML sanitization.

    Raises:
        TypeError: If HTML is not a string
    """
    if not isinstance(html, str):
        raise TypeError(ValidationErrorMessages.HTML_MUST_BE_STRING)

    if allowed_tags is None:
        # Remove all HTML tags
        sanitized = re.sub(r"<[^>]+>", "", html)
    else:
        # Remove all tags except allowed ones
        # This is a simplified implementation
        pattern = "|".join(allowed_tags)
        sanitized = re.sub(rf"<(?!\/?(?:{pattern})\b)[^>]+>", "", html, flags=re.IGNORECASE)

    return sanitized.strip()
